### ISSUE TITLE
Allow migration of followers to new instance when old instance is unresponsive

### ISSUE THREAD

**Issue by lapcat** [mastodon/mastodon#23401]
### Pitch

There are a number of possible implementations. One is that users have the ability at any time to export their list of followers, just like they can export their follows, and the instance cryptographically signs the list of followers on export. Good third-party clients could even automatically export and store the followers list for the convenience of users. Then the user could import the cryptographically signed list on their new instance, and the new instance could verify the list with the old instance's public key.

### Motivation

I've written about how the mstdn.plus instance suddenly broke this week, and the administrator has gone missing, thereby stranding 4500 users, leaving them unable to move their followers to new instances: https://lapcatsoftware.com/articles/mastodon.html The same person also administrates (or used to administrate) the mstdn.party instance with 12,000 users, so this kind of situation is likely to happen again, perhaps soon. There needs to be a way for users to move their followers to new instances without the cooperation of the old instance.

**Comment by simon-frankau** [comment]
As an SRE, the sudden, complete disappearance of an instance worries me. Small instances aren't exactly selling themselves on the effectiveness of their back-up regimes, and a scheme where users don't have to trust their admin to maintain their accounts seems a very good thing.

Guessing at implementations, a signed recovery doc sounds like the way to me. It could be argued that leaves more pathways to abuse than the regular hand-over protocol, but I'm not exactly sure exactly how well-hardened the existing infrastructure is against e.g. an incompliant instance allowing an account to be transferred out to multiple destinations.

**Comment by klrtk** [comment]
I think maybe doing it the other way around would be a better option, have all the clients create their own keypairs and the instance distribute the public keys to followers for caching. It should take several days for a key to become trusted. A move could be initiated by contacting all followers' instances, it should also be reversible using the same keys during a grace period, in which case they also become revoked.

This would have several advantages:
- Users could migrate if the instance goes down
- An instance compromise wouldn't put user accounts at risk of being irreversibly migrated
- A user's device being compromised would still allow the user to recover their account
- Could potentially guard against an admin who suddenly becomes uncooperative and attempts to block moves out of the instance

Some other changes that could enhance this:
- Defederation shouldn't completely sever follower relationships and be reversible, moving to a different instance should restore suspended followers
- The owner of a suspended account should be able to migrate against the instance's wish, this could only be allowed with special approval of the receiving instance to make abuse harder

I think this is really important because instances run by individuals and small organizations can be less reliable, and personal conflict can affect users suddenly without giving them a chance to make a choice, as we've seen with the latest situation on mstdn.social.

**Comment by ChlorideCull** [comment]
It's worth noting that you don't need to add another layer of keys and cryptography on top of this - ActivityPub already gives each user a unique key ([see `publicKeyPem` on my user for example](https://masto.fuzzy.systems/@ChlorideCull.json)) which is already used to verify that a post comes from the user it claims to come from - all we need is for that key to be exportable, and some way to "gossip" migration messages between instances, where we don't care where the message came from, just that it's valid.

**Comment by trwnh** [comment]
> ActivityPub already gives each user a unique key ([see `publicKey` on my user for example](https://masto.fuzzy.systems/@ChlorideCull.json)) which is already used to verify that a post comes from the user it claims to come from - all we need is for that key to be exportable, and some way to "gossip" migration messages between instances, where we don't care where the message came from, just that it's valid.

this also requires fundamentally redefining identity to be rooted in the key, and not the actual id*. the strength and weakness of ActivityPub is that it is part of the Web.

* tbh even the id is not canonical in mastodon, mastodon identifies users internally throughout the codebase by the username+domain pair rather than their canonical id

one other thing to point out is that `publicKey` and its usage is something that comes from the [Security Vocabulary](https://w3id.org/security#publicKey) and was bolted on by mastodon in order to do HTTP Signatures. it's not strictly part of ActivityPub. that's more of a technicality though

**Comment by ChlorideCull** [comment]
> this also requires fundamentally redefining identity to be rooted in the key, and not the actual id*. the strength and weakness of ActivityPub is that it is part of the Web.
> 
> 
> 
> * tbh even the id is not canonical in mastodon, mastodon identifies users internally throughout the codebase by the username+domain pair rather than their canonical id

Not necessarily - we could just fall back to it in case we can't access the authoritive server.

I'm personally a bit annoyed that the spec doesn't touch on how to handle the authoritive server being down, and the whole "how exactly should we handle authentication and authorization" part hadn't shaken out yet.

The best guidance the spec offers is B.2 - which says

>Servers should not trust client submitted content, and federated servers also should not trust content received from a server other than the content's origin without some form of verification.

This does imply that it's okay to trust content from a non-authorative server, as long as there's been some sort of verification. Having the objects be signed by the actor key would be good enough for me to consider it verification.

> one other thing to point out is that `publicKey` and its usage is something that comes from the [Security Vocabulary](https://w3id.org/security#publicKey) and was bolted on by mastodon in order to do HTTP Signatures. it's not strictly part of ActivityPub. that's more of a technicality though

Fair. Missed that. Seems to have become the accepted implementation for authentication and authorization, though.

**Comment by klrtk** [comment]
> It's worth noting that you don't need to add another layer of keys and cryptography on top of this - ActivityPub already gives each user a unique key (see publicKey on my user for example) which is already used to verify that a post comes from the user it claims to come from - all we need is for that key to be exportable, and some way to "gossip" migration messages between instances, where we don't care where the message came from, just that it's valid.

I disagree, I don't think actor keys provide good security for this as they're long-term and controlled by the instance. The goal I have in mind is to prevent an instance from bringing a user account down completely in case that user has somewhere to move to, this includes instance compromises, sudden shutdowns or even malicious admins.

I know there are other roadblocks here, like the way self-destruct and suspensions work, which warrants its own issue, but designing this system to accommodate it right away would be better, because it requires protocol changes anyway, which is only getting harder as the network grows and new implementations are added.

**Comment by FedericoCeratto** [comment]
Being able to upload and download the account keypair could be used for trusted account migration. As an added benefit signing could be done entirely on desktop/mobile clients, removing the risk of account impersonation when an instance is compromised. 
Partially related: #928 (GPG signature verification).

**Comment by therealbnut** [comment]
If there’s protocol level support for distributing per-user public keys, which could be used for server migration, perhaps they could also be used for end-to-end encrypted DMs (which seems like it would address a common trust issue).

I’m not sure how your proposal would handle some often requested things like:
 * migrating posts
 * migrating DMs
 * migrating media
 
I’m sure a user could create a signed archive of  their content, but I’m not sure how it could be verified. It seems like an avenue for abuse for a user to be able to rewrite history, even if it’s their own.

Maybe you can designate a server as a mirror, and any ActivityPub messages go to both servers from the start, cross verifying.

—

Either way, perhaps this migration issue could make progress if it starts with just the protocol change for distributing/associating keys and encrypting objects. That will likely be useful in many other ways too.

The protocol changes would ideally be pretty generic to allow features to be added later, ie. new object and actor fields in ActivityPub to support object content encrypted or signed by a designated key owner.

[libSignal](https://github.com/signalapp/libsignal) may be a good place to start when considering how to approach the protocol changes.
